README
This package contains the source for the function strasp() (in strasp.c)
which seperates (or "splits") a string according to a string of delimeters
and returns an array of pointers to *newly* allocated strings (i.e., char **).

(strasp stands for "string array seperator")

The motivation for this function is to remove string manupulation that occurs
with functions like `strsep'; the original string sent to strasp() (i.e.,
`str') will not be modified in any way. However, the draw back to this is
that the newly allocated array needs to be free'd().

The convenience function strasp_free() is supplied (defined in strasp.c) to
free the array of char's.

For example,

--
char **array;
char **t_array;

array = strasp("PATH=/usr/bin:/sbin:/usr/local/bin", ":");

if (array == NULL)
	exit(1);

/* move another pointer along array (otherwise, free will croak) */
t_array = array;
while (t_array && *t_array)
	printf("%s\n", *(t_array++));

strasp_free(array);
--

The above code segment will print the following:

PATH=/usr/bin
/sbin
/usr/local/bin

This package also contains a shell-like driver for testing strasp(). The
file `main.c' is compiled (with `strasp.c') into the binary `strasp'
by simplying calling the command `make' after configuring (See INSTALL).

To use `strasp', invoke it by sending the delimeter as the first argument and
each "command" at the strasp prompt (i.e., "strasp>") is a string you want to
split up. For example,

$ ./strasp ':'
strasp> PATH=/usr/bin:/sbin:/usr/local/bin
PATH=/usr/bin
/sbin
/usr/local/bin
strasp> i:like:colon:seperated:sentnces.
i
like
colon
seperated
sentnces.



ChangeLog
2004-01-17 Erick Bourgeois <erick@jeb.ca>
	* configure.in: Updated version 1.8
	* strasp.c (__strasp_trim_delims): Removed this function so that
	malloc is used as little as possible.
	* strasp.c (strasp): Function was considerably changed to accomadate
	for the removal of the above function (i.e. `__strasp_trim_delims').
	Trimming is now done at the beginning of this function to reducde the
	number of passes over the passed string (i.e. `str').
	Now it runs over `str' only once, thus O(n).
	
2004-01-04 Erick Bourgeois <erick@jeb.ca>
	* strasp.h: Removed RETURN_ARRAY macro to reduce ambiguity.

2003-12-26 Erick Bourgeois <erick@jeb.ca>
	* configure.in: Changed STRASP_BUFSIZ template and "arg with" to
	STRASP_ELEBUFSIZ, so that the element buffersize can be set at compile
	time.
	Created STRASP_ARRBUFSIZ template and "arg with" argument for
	`configure'.
	* strasp.h: Removed STRASP_BUFSIZ and made STRASP_ELEBUFSIZ be defined
	directly to BUFSIZ.
	Made STRASP_ARRBUFSIZ be a right logical shift 4 of STRASP_BUFSIZ,
	that is, STRASP_BUFSIZ / 16

2003-12-25 Erick Bourgeois <erick@jeb.ca>
	* strasp.c (__strasp_xmalloc): Removed upper bound restriction of
	STRASP_ELEBUFSIZ.
	Assigned `ptr' to the return value of `memcpy'.
	* strasp.c (__strasp_trim_delims): Renamed `end' variable to `tail'.
	Renamed `t_str' variable to `head'.
	Adjusted comments.
	* strasp.c (strasp): Completly changed the algorithm:
	Instead of passing over the array for the number of seperate strings
	and then passing over it again to extract the strings, I implemented
	a one pass excluding the triming phase, which, in worst case, be
	O(n-1), i.e., all char's of delim with one char not in delim in the
	middle of str. The one pass dynamically allocates memory for the array
	with a new macro, STRASP_ADDELEM (see below).
	
	* strasp.h: Added RETURN_ARRAY, STRASP_ADDELEM, and STRASP_INSERT_STR.
	o RETURN_ARRAY: macro for free'ing the trimmed str, NULL
	  terminating the array and returning the array.
	o STRASP_ADDELEM: macro for testing the possibility of adding an
	element to the array and then realloc'ing the space for another
	string.
	o STRASP_INSERT_STR: macro for testing the possibility of adding a
	string to the last allocated element (i.e., it should follow a 
	STRASP_ADDELEM) and then allocating and assinging it to the second to
	last element (i.e., a_size - 2) in `array' (since a_size - 1 is for
	'\0').
	
2003-12-22 Erick Bourgeois <erick@jeb.ca>
	* configure.in: Added yes/no checking for ARG_WITH option 
	strasp-bufsiz.
	Removed double checks for the already checked (with AC_HEADER_STDC
	macro) STDC headers.
	
2003-12-20 Erick Bourgeois <erick@jeb.ca>
	* GLOBAL: Renamed the file strasp.main.c to main.c.
	* configure.in: Added warnings for missing headers: stdio.h, stdlib.h,
	and string.h
	* strasp.c: Added more comments to (and in) strasp().
	* strasp.c (__strasp_xmalloc): Testing buffer length now incorporates
	len of 0, instead of just < 0.
	* strasp.c (__strasp_trim_delims): Added more comments.
	* main.c: Made string the size of STRASP_BUFSIZ instead of BUFSIZ.
	* Makefile.am: Changed strasp_SOURCES to incorporate the file cahnge
	mentioned above.

2003-12-17  Erick Bourgeois <erick@jeb.ca>
	* strasp.c: Added RCSID
	* strasp.h: Added `configure.in' and `Makefile.am'.
	Added definition checks for BUFSIZ and STRASP_BUFSIZ:
	if set with `configure', then that value will be used.
	* strasp.main.c: Added RCSID
	* strasp.main.c (main):	Removed mallinfo() 
	function that was used to monitor malloc calls.
	Added a shell like interface and removed static example.

2003-02-17  Erick Bourgeois <erick@jeb.ca>
	* strasp was created


main.c
/* Copyright 2003
 * Erick Bourgeois, <erick@jeb.ca>
 * This is free software distributed under the terms of the
 * GNU Public License.
 * 
 * This shell like driver is used to demonstrate the strasp function.
 * 
 */

#if HAVE_CONFIG_H
#  include <config.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "strasp.h"

static char *rcsid = "$Id: main.c,v 1.2 2003/02/15 14:13:34 EST ebourgeois Exp $";

int
main (int argc, char **argv)
{
	char string[STRASP_ARRBUFSIZ];
	char *str_ptr = string;
	const char *delim = "";
	char **myargv;
	char **tmp_array;
	
	if (argc != 2)
	{
		printf("usage: strasp \"delim\"\n");
		return 1;
	}
	delim = argv[1];
	
	while (1)
	{
		printf("strasp> ");
		if ((str_ptr = fgets(str_ptr, STRASP_ELEBUFSIZ, stdin)) == NULL)
		{
			perror("gets");
			return 1;
		}
		/* split the command line into `words',
		 * delimited by any char in delim */
		str_ptr[strlen(str_ptr)-1] = '\0';
		myargv = strasp(str_ptr, delim);
		
		if (myargv == NULL)
			continue;

		tmp_array = myargv;
		while (tmp_array && *tmp_array)
			printf("%s\n", *(tmp_array++));

		strasp_free(myargv);
	}
	
	return 0;
}



strasp.c
/* Copyright 2003
 * Erick Bourgeois, <erick@jeb.ca>
 * This is free software distributed under the terms of the
 * GNU Public License.
 * 
 * These 4 functions (2 of which are static) comprise strasp.
 * strasp() seperates, or splits, a string according to a string
 * of delimeters and returns an array of pointers to newly
 * allocated strings.
 * 
 * The motivation for this function is to remove string
 * manupulation that occurs with strsep; the original string
 * sent to strasp() (i.e., `str') will not be modified in any
 * way. However, the draw back to this is that the newly
 * allocated array needs to be free'd() (See below).
 * 
 * Note: the array needs to be free'd(). The non-static 
 * strasp_free() function is provided for this reason.
 */

#if HAVE_CONFIG_H
#  include <config.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "strasp.h"

static char *rcsid = "$Id: strasp.c,v 1.8 2003/02/17 1:13:34 EST ebourgeois Exp $";
#if 0
#if __GNUC__ == 2
#define USE(var) static void * use_##var = (&use_##var, (void *) &var) 
USE (rcsid);
#endif
#endif

/* This function safely allocates len + 1 bytes
 * and copies len bytes starting from ptr, then
 * it NULL terminates the new string.
 *
 * Precondition: none
 * Return value: a pointer to the NULL terminated 
 *               allocated memory, otherwise NULL.
 *               That is, if len <= 0, or if it 
 *               could not allocate enough memory.
 */
static char *
__strasp_xmalloc(char *str, int len)
{
	char *ptr;
	
	if (len <= 0)
                return NULL;
       
        ptr = (char *) malloc(sizeof(char) * (len + 1));
        if (ptr == NULL)
                return NULL;
	
        ptr = memcpy(ptr, str, len);
        ptr[len] = '\0';
        
        return ptr;
}

/* This function free's() an **array. It is needed if something
 * may go wrong internally and we have already mallocated
 * **array.
 *
 * Precondition: none
 * Postcondition: array is now NULL
 * Return value: none
 */
void
strasp_free(char **array)
{
        unsigned int i;
        
        if (array == NULL)
                return;
        
        for (i = 0; array[i] != NULL; i++)
                free(array[i]);
       
        free(array);
        
        return;
}

/* A general string splitter.
 * Given a string, split it up according to any
 * character in delim, then return an array of
 * these values.
 * 
 * Precondition: none
 * Return value: char ** (array of char *'s)
 * Note: return value should be free'd()
 */
char **
strasp(char *str, const char *delim)
{
	char *head;
	char *tail;
	char *tmp;
	char **array;
	/* array size */
	size_t a_size;
	size_t diff;
	
	if (str == NULL || delim == NULL || strlen(str) == 0)
		return NULL;
		
	head = str;
	/* remove all occurences of any char in delim
	   from the start and the end of str */
	while (*head != '\0' && (strchr(delim, *head++) != NULL))
                ;
        
	head--;
        /* nothing is left in head */
        if (head > str)
                if (strlen(head) == 0)
                        return NULL;

	tmp = head + strlen(head);
        /* remove all occurences of any char in delim
	   from the start and the end of str */
	while (strchr(delim, *--tmp) != NULL)
                ;
        
        /* nothing is left at the end */
        if (head >= ++tmp)
                return NULL;

	/* initiallize */
	tail = tmp;
	tmp = head;
	array = NULL;
	
	/* 1 element since strlen(str) > 0
	   and the other for '\0' */
        a_size = 2;
	
        /* let's get to it; process the supplied string
           how many characters in delim occur in str. */
	while (tmp != tail)
	{		
		if (strchr(delim, *tmp) != NULL)
		{
			diff = tmp - head;
			
			STRASP_ADDELEM(array, a_size);
			STRASP_INSERT_STR(array, a_size, head, diff);
			
			/* move tail along while it is in delim */
			while (*tmp != '\0' && strchr(delim, *++tmp) != NULL)
				;
			
			head = tmp;
			a_size++;
		}
		tmp++;

		if (*tmp == '\0' || tmp == tail)
		{
			diff = tmp - head;
			
			STRASP_ADDELEM(array, a_size);
			STRASP_INSERT_STR(array, a_size, head, diff);
			
			a_size++;
		}
	}

	if (array != NULL || a_size != 2)
		array[a_size - 2] = '\0';
	
	char **tmp_a = array;
	
	return array;
}



strasp.h
/* Copyright (C) 2000
 * Erick Bourgeois, <erick@jeb.ca>
 * This is free software distributed under the terms of the
 * GNU Public License.
 *
 * $Id: strasp.h,v 1.8 2003/02/17 1:13:34 ebourgeois Exp $
 */

#ifndef _STRASP_H
#define _STRASP_H

/* Define a suitable buffer size */
#ifndef BUFSIZ
#  define BUFSIZ 8192
#endif

/* Each element in the array returned by strasp is a maximum
   length of STRASP_ELEBUFSIZ */
#ifndef STRASP_ELEBUFSIZ
#  define STRASP_ELEBUFSIZ BUFSIZ
#endif

/* By default, the array holds 512 pointers/elements */
#ifndef STRASP_ARRBUFSIZ
#  define STRASP_ARRBUFSIZ (STRASP_ELEBUFSIZ >> 4)
#endif

#define __STRASP_ADDELEM(array, a_size) \
	if (a_size > STRASP_ARRBUFSIZ) \
	{ \
		array[a_size - 2] = '\0'; \
		return array; \
	} \
	array = (char **) realloc(array, a_size * sizeof(char *)); \
	if (array == NULL) \
	{ \
		array[a_size - 2] = '\0'; \
		return array; \
	}
#define STRASP_ADDELEM __STRASP_ADDELEM
	
#define __STRASP_INSERT_STR(array, a_size, str, len) \
	if (len > STRASP_ELEBUFSIZ) \
	{ \
		array[a_size - 2] = '\0'; \
		return array; \
	} \
	array[a_size - 2] = __strasp_xmalloc(str, len); \
	if (array[a_size - 2] == NULL) \
	{ \
		array[a_size - 2] = '\0'; \
		return array; \
	}
#define STRASP_INSERT_STR __STRASP_INSERT_STR
	
/* Provide this as a convenience free'n function */
void strasp_free (char **array);
char **strasp (char *str, const char *delim);

#endif /* !_STRASP_H */

